src/tactic/compute_degree.lean

```diff
@@ -172,8 +298,182 @@ by compute_degree_le
+  resolve_coeff
+| `(%%a * %%b) := do
+  [da, db] ← [a,b].mmap guess_degree,
+  refine ``((coeff_mul_of_nat_degree_le' (by norm_num : %%da + %%db = %%n) _ _).trans _),
+  iterate_at_most' 2 compute_degree_le,
+  try $ congr' (some 1);
+  resolve_coeff
+| `(%%a ^ %%ex) := do
+  da ← guess_degree a,
+  refine ``((coeff_pow_of_nat_degree_le' (by norm_num : %%da * %%ex = _) _).trans _),
+  compute_degree_le,
+  try $ congr' (some 1);
+  resolve_coeff
+| `(- %%a) := do
+  refine ``((coeff_neg _ _).trans (neg_inj.mpr _)),
+  resolve_coeff
+| e := skip
+end
+
+namespace interactive
+open compute_degree
+
+/--
+`simp_coeff` assumes that the target is either of the form `f.coeff n = x` or of the form
+`f.coeff n ≠ x`.  It then proceeds to simplify `f.coeff n` recurring to the pair
+`get_lead_coeff/resolve_coeff` to scan the expression of `f` and producing a hopefully simpler
+expression.  After this, it calls on `simp only [a few lemmas]` and `norm_num` to simplify further.
+-/
+meta def simp_coeff : tactic unit :=
+do t ← target >>= instantiate_mvars,
+  (t_is_eq, f, m) ← match t with
+    | `(coeff %%f %%m = _) := return (tt, f, m)
+    | `(coeff %%f %%m ≠ _) := return (ff, f, m)
+    | _ := fail "Goal is not of the form `f.coeff n = x` or `f.coeff n ≠ x`"
+    end,
+  d_nat ← guess_degree' f,
+  m_nat ← eval_expr' ℕ m,
+  guard (d_nat = m_nat) <|> fail!(
+  "`simp_coeff` checks that the expected degree is equal to the degree appearing in `coeff`\n" ++
+  "the expected degree is `{d_nat}`, but you are asking about the coefficient of degree `{m_nat}`"),
+  `(@polynomial %%R %%_) ← infer_type f,
+  c ← mk_instance_cache R,
+  (c, lc) ← get_lead_coeff c f,
+  nn ← get_unused_name "c_c",
+  cf ← to_expr ``(coeff : polynomial %%R → ℕ → %%R),
+  co_eq_co ← mk_app `eq [cf.mk_app [f, m], lc],
+  c_c ← assert nn co_eq_co,
+  interactive.swap,
+  if t_is_eq then refine ``(eq.trans %%c_c _) else refine ``(ne_of_eq_of_ne %%c_c _),
+  try $ tactic.clear c_c,
+  interactive.swap,
+  resolve_coeff
+
+section parsing
+setup_tactic_parser
+
+/--
+`reduce_coeff` takes a polynomial `f` as an input and produces a hypothesis of the form
+`c_c : f.coeff n = <simplified expression>`, where the simplified expression is obtained via the
+`get_lead_coeff/resolve_coeff` pair.
+-/
+meta def reduce_coeff (fp : parse texpr) : tactic unit :=
+do
+  f ← to_expr ``(%%fp) tt ff,
+  exp_deg ← guess_degree' f,
+  `(@polynomial %%R %%_) ← infer_type f,
+  c ← mk_instance_cache R,
+  (c, lc) ← get_lead_coeff c f,
+  nn ← get_unused_name "c_c",
+  cf ← to_expr ``(coeff : polynomial %%R → ℕ → %%R),
+  co_eq_co ← mk_app `eq [cf.mk_app [f, `(exp_deg)], lc],
+  assert nn co_eq_co,
+  resolve_coeff
+
+end parsing
+
+/--  `compute_degree` tries to close goals of the form `f.(nat_)degree = d`.  It converts the
+goal to showing that
+* the degree is at most `d`, calling `compute_degree_le` to solve this case;
+* the coefficient of degree `n` is non-zero, calling `simp_coeff` to simplify this goal.
```

Comment:
I'd like the docstring to mention that we have an expected degree.
```suggestion
* there is an upper bound `n` for `f.nat_degree`, assuming no terms cancel;
* `n` and `d` coincide;
* the degree is at most `d`, calling `compute_degree_le` to solve this case;
* the coefficient of degree `d` is non-zero, calling `simp_coeff` to simplify this goal.
```
Or is this too much detail?

Reply:
I actually had not realized that I used `d` and `n`: they should have been the same!  I also reworded slightly the rest, so maybe now this comment is settled?